File: valid_parenthesis.py
# Time Complexity :O(N)
# Space Complexity :O(N)
# Did this code successfully run on Leetcode : Yes

# Any problem you faced while coding this :

# Your code here along with comments explaining your approach

# stack solution for many types of parenthesis
from collections import deque


class Solution:
    def isValid(self, s: str) -> bool:
        l=deque()
        for i in s:
            if( i in ['{','[','(']):
                l.append(i)
            else:
                if len(l)==0:
                    return False
                curr=l.pop()
                if(curr=='(' and i !=')'):
                    return False
                if(curr=='[' and i!=']'):
                    return False
                if(curr=='{' and i!='}'):
                    return False
        if(len(l)):
            return False
        return True

# just use a count variable if there are same type of brackets.

File: sol/ValidParenthesis.java
package sol;

import java.util.Stack;

public class ValidParenthesis {
// Time complexity :O(n)
//	Space: O(n)
	  public boolean check(char a, char b){
	        
	        if(a=='{'&&b=='}'){
	            return true;
	        }else if (a=='(' && b==')'){
	            return true;
	        }else if (a=='['&&b==']'){
	            return true;
	        }
	        return false;
	    }
	    
	    
	    
	    public boolean isValid(String s) {
	        
	        
	        Stack<Character> st= new Stack <Character>();
	        int p=0;
	        
	        while(p<s.length()){
	            
	        if(st.isEmpty()){
	            
	            st.push(s.charAt(p));
	            p++;
	        }else if(!check(st.peek(),s.charAt(p))) {
	            
	            st.push(s.charAt(p));
	            p++;
	            
	        }else if(check(st.peek(),s.charAt(p))){
	            st.pop();
	            p++;   
	                   }
	            
	            
	        }
	        
	        
	        if(st.isEmpty()){return true;}else{return false;}
	        
	    }
}
